### PR TITLE
feat: add email to A/B function

### LINKED ISSUE #39125: Enabling A/B testing
It would be helpful to have different versions of the site live so we could make more data-driven decisions regarding the changes that we introduce to the the client.

### PR DESCRIPTION
**Pull request by ahmaxed**
This PR focuses on setting up A/B testing for singed in users when Donation related GA events are executed.

I uses emails of the signed in users as a unique identifier, hash it, turn it to a number, and assign A or B variation if the number is even or not.

This PR is intended to land on the .dev to see if the related custom dimensions ( event's test variation, etc.) are being reported to google analytics. 

Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #39125

### CODE REVIEW

**Comment by ahmadabdolsaheb** [comment]
@ojeytonwilliams, could you take a quick look at the function to see if it makes sense to you. There are some tiny TS stuff that I would appreciate you taking a look at as well.

File: client/src/utils/A-B-tester.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+} {
+  // turn the email into a number
+  const hash: string = sha1(email);
+  const hashInt = parseInt(hash, 16);
+
+  // turn the number to A or B variant
+  const maxhash = 16 ** 40;
+  const maxDivisor = maxhash / 100;
+  const simplifiedInt = Number((hashInt / maxDivisor).toFixed());
+  const isAVriant = simplifiedInt % 2 === 0;
```
Comment (ojeytonwilliams) [inline client/src/utils/A-B-tester.ts:19]:
```suggestion
  const hashInt = parseInt(hash.slice(0,3), 16);

  // turn the number to A or B variant
  const isAVriant = hashInt % 2 === 0;
```
How about this?

**Comment by ahmadabdolsaheb** [comment]
@ojeytonwilliams round 2?

File: client/src/redux/ga-saga.test.js [hunk 2]
```diff
@@ -11,8 +11,8 @@ describe('ga-saga', () => {
     const mockEventPayload = {
       type: 'event',
       data: {
-        category: 'Donation',
-        action: 'year end gift paypal button click'
+        category: 'Map Challenge Click',
+        action: '/learn'
```
Comment (ahmadabdolsaheb) [inline client/src/redux/ga-saga.test.js:15]: Changed the mock data to reduce the debugging time before the feature is complete.

File: client/src/redux/ga-saga.js [hunk 3]
```diff
@@ -1,9 +1,39 @@
...
 
 const GaTypes = { event: ga.event, page: ga.pageview, modal: ga.modalview };
 
 function* callGaType({ payload: { type, data } }) {
+  if (
+    type === 'event' &&
+    data.category.includes('Donation') &&
+    ABTestConfig.isTesting
+  ) {
+    const isSignedIn = yield select(isSignedInSelector);
```
Comment (ahmadabdolsaheb) [inline client/src/redux/ga-saga.js:21]: yield only after all the other conditions are met.

File: client/src/utils/sha-1.d.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+export declare module 'sha-1';
```
Comment (ojeytonwilliams) [inline client/src/utils/sha-1.d.ts:1]: I don't think this is needed any more.

File: client/src/utils/A-B-tester.test.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    };
+    const result = emailToAVariant('foo@freecodecamp.org');
+    expect(result).toEqual(preSavedResult);
+  });
+  it('Distributes A and B variations equaly for 100K random emails', () => {
+    let A = 0;
+    let B = 0;
+    const sampleSize = 100000;
+    for (let i = 0; i < sampleSize; i++) {
+      if (emailToAVariant(faker.internet.email()).isAVriant) A++;
```
Comment (ojeytonwilliams) [inline client/src/utils/A-B-tester.test.ts:19]: Can we seed faker?  We don't need to keep checking to see if different samples fall close to 50/50.  

I know the chance is low, but I don't think we gain anything by allowing this the possibility of failure.

File: client/src/utils/A-B-tester.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import sha1 from 'sha-1';
+
+// This function turns an email to a hash and decides if it should be
+// an A or B variant for A/B testing
+
+export function emailToAVariant(email: string): {
+  hash: string;
+  isAVriant: boolean;
```
Comment (ojeytonwilliams) [inline client/src/utils/A-B-tester.ts:8]:
```suggestion
  isAVariant: boolean;
```
Typo?

File: client/src/utils/A-B-tester.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  hashInt: number;
+} {
+  // turn the email into a number
+  const hash: string = sha1(email);
+  const hashInt = parseInt(hash.slice(0, 1), 16);
+  // turn the number to A or B variant
+  const isAVriant = hashInt % 2 === 0;
```
Comment (ojeytonwilliams) [inline client/src/utils/A-B-tester.ts:15]:
```suggestion
  const isAVariant = hashInt % 2 === 0;
```

File: client/src/utils/A-B-tester.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  return {
+    hash,
+    isAVriant,
```
Comment (ojeytonwilliams) [inline client/src/utils/A-B-tester.ts:18]:
```suggestion
    isAVariant,
```

File: client/src/redux/ga-saga.js [hunk 9]
```diff
@@ -1,9 +1,39 @@
...
+    if (isSignedIn) {
+      const email = yield select(emailSelector);
+      const completedChallengeTotal = yield select(completedChallengesSelector);
+      const completedChallengeSession = yield select(completionCountSelector);
+      const customDimentions = {
```
Comment (ojeytonwilliams) [inline client/src/redux/ga-saga.js:26]:
```suggestion
      const customDimensions = {
```

File: client/src/redux/ga-saga.js [hunk 10]
```diff
@@ -1,9 +1,39 @@
...
+        Test_Variation: emailToAVariant(email) ? 'A' : 'B',
+        Test_Type: ABTestConfig.type,
+        Challenges_Completed_Session: completedChallengeSession,
+        Challenges_Completed_Total: completedChallengeTotal.length,
+        URL: window.location.href
+      };
+      data = { ...data, ...customDimentions };
```
Comment (ojeytonwilliams) [inline client/src/redux/ga-saga.js:33]:
```suggestion
      data = { ...data, ...customDimensions };
```

File: client/src/utils/A-B-tester.test.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
+import faker from 'faker';
+import { emailToAVariant } from './A-B-tester';
+
+describe('client/src is-email-variation-a', () => {
+  it('Consistently returns the same result for the same input', () => {
+    const preSavedResult = {
+      hash: '23e3cacb302b0c759531faa8b414b23709c26029',
+      isAVriant: true,
```
Comment (ojeytonwilliams) [inline client/src/utils/A-B-tester.test.ts:8]:
```suggestion
      isAVariant: true,
```

File: client/src/redux/ga-saga.js [hunk 12]
```diff
@@ -1,9 +1,39 @@
-import { takeEvery, call, all } from 'redux-saga/effects';
+/* eslint-disable camelcase */
+import { takeEvery, call, all, select } from 'redux-saga/effects';
+import { ABTestConfig } from '../../../config/donation-settings';
```
Comment (ojeytonwilliams) [inline client/src/redux/ga-saga.js:3]:
```suggestion
import { aBTestConfig } from '../../../config/donation-settings';
```
Nitpicky, so feel free to ignore

File: client/src/redux/ga-saga.js [hunk 13]
```diff
@@ -1,9 +1,39 @@
...
 import ga from '../analytics';
+import {
+  isSignedInSelector,
+  emailSelector,
+  completionCountSelector,
+  completedChallengesSelector
+} from '../redux';
+import { emailToAVariant } from '../utils/A-B-tester';
```
Comment (ojeytonwilliams) [inline client/src/redux/ga-saga.js:11]:
```suggestion
import { emailToABVariant } from '../utils/A-B-tester';
```
The first one reads like "email to a variant", but this reads "email to AB variant", at least to me.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: client/src/utils/sha-1.d.ts (hunk 4 above)
Comment (ahmadabdolsaheb) [reply client/src/utils/sha-1.d.ts:1]: Removed and TS server started crying, I will give it another try.

File: client/src/utils/A-B-tester.test.ts (hunk 5 above)
Comment (ahmadabdolsaheb) [reply client/src/utils/A-B-tester.test.ts:19]: I will look into seeding. If we don't test the 50/50 a/b distribution, what should we test?

Comment (ojeytonwilliams) [reply client/src/utils/A-B-tester.test.ts:19]: It's fine to test the distribution, but it should be deterministic.  If we call `faker.seed(123);` or whatever seed you like, then the tests will always generate the same ratio.

File: client/src/utils/sha-1.d.ts (hunk 4 above)
Comment (ojeytonwilliams) [reply client/src/utils/sha-1.d.ts:1]: I'll take a look.

**Review by ojeytonwilliams — DISMISSED** [review]
LGTM :+1:

**Review by ojeytonwilliams — DISMISSED** [review]

File: client/src/redux/ga-saga.js [hunk 14]
```diff
@@ -1,9 +1,39 @@
...
+    data.category.includes('Donation') &&
+    aBTestConfig.isTesting
+  ) {
+    const isSignedIn = yield select(isSignedInSelector);
+    if (isSignedIn) {
+      const email = yield select(emailSelector);
+      const completedChallengeTotal = yield select(completedChallengesSelector);
+      const completedChallengeSession = yield select(completionCountSelector);
+      const customDimensions = {
+        Test_Variation: emailToABVariant(email) ? 'A' : 'B',
```
Comment (ShaunSHamilton) [inline client/src/redux/ga-saga.js:27]:
```suggestion
        Test_Variation: emailToABVariant(email).isAVariant ? 'A' : 'B',
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
If I am not mistaken, without this, the `Test_Variation` will always be `'A'`

**Review by ShaunSHamilton — APPROVED** [review]
This LGTM 👍

**Review by ojeytonwilliams — APPROVED** [review]
LGTM 👍

**Pull request merged**